### PR TITLE
[RHCLOUD-35647] Lock group for adding principals and deletion

### PR DESCRIPTION
**Pull request by lpichler**
## Link(s) to Jira
- https://issues.redhat.com/browse/RHCLOUD-35647

## Description of Intent of Change(s)
`get_queryset` method was changed to lock groups for mentioned actions.
`get_group_queryset` method  retuns query set which contains GROUP BY and this statement cannot be  combined together with FOR UPDATE statement.

### Required code code change - to wait between deleting group
```
@@ -392,7 +392,8 @@ class GroupViewSet(

             dual_write_handler = RelationApiDualWriteGroupHandler(group, ReplicationEventType.DELETE_GROUP)
             dual_write_handler.prepare_to_delete_group()
-
+            import time
+            time.sleep(5)
             response = super().destroy(request=request, args=args, kwargs=kwargs)

             dual_write_handler.replicate()
```

Group is needed for test case.

### Test case 
1. call delete group  API endpoint
```DELETE /api/rbac/v1/groups/8010bbdd-6d2a-40d9-a30a-4866fb68dcef/
```
This call is blocked by `time.sleep(5)`

2.call add principals to group API endpoint sooner than `time.sleep(5)` expires

```
POST /api/rbac/v1/groups/8010bbdd-6d2a-40d9-a30a-4866fb68dcef/principals/
```

### Results

Expected behaviour: Second call fails with 404 Not found status code  - this call waits on completion of first call and when it resulted successfully it removed group, so it will be basically adding principals api endpoint on group which doesn't exist == 404 status code.

Behaviour before:  Second call was successful, write produces relations and those relations were not removed by first call.

## Checklist
- [ ] if API spec changes are required, is the spec updated?
- [ ] are there any pre/post merge actions required? if so, document here.
- [ ] are theses changes covered by unit tests?
- [ ] if warranted, are documentation changes accounted for?
- [ ] does this require migration changes?
  - [ ] if yes, are they backwards compatible?
- [ ] is there known, direct impact to dependent teams/components?
  - [ ] if yes, how will this be handled?

## Secure Coding Practices Checklist Link
- https://github.com/RedHatInsights/secure-coding-checklist

## Secure Coding Practices Checklist
- [ ] Input Validation
- [ ] Output Encoding
- [ ] Authentication and Password Management
- [ ] Session Management
- [ ] Access Control
- [ ] Cryptographic Practices
- [ ] Error Handling and Logging
- [ ] Data Protection
- [ ] Communication Security
- [ ] System Configuration
- [ ] Database Security
- [ ] File Management
- [ ] Memory Management
- [ ] General Coding Practices

### CODE REVIEW

File: rbac/management/group/view.py [hunk 1]
```diff
@@ -681,7 +690,9 @@ def principals(self, request: Request, uuid: Optional[UUID] = None):
                 else:
                     principals.append(specified_principal)
 
-            self.protect_system_groups("add principals", group)
+            with transaction.atomic():
+                group = self.get_object()
+                self.protect_system_groups("add principals", group)
 
             if not request.user.admin:
```
Comment (alechenninger) [inline rbac/management/group/view.py:697]: TODO: Move everything under here under the same atomic block

Comment (lpichler) [reply rbac/management/group/view.py]: Done 👍

**Review by alechenninger — APPROVED** [review]
LGTM

**Comment by alechenninger** [comment]
Btw we might be able to test for this repeatedly, check out `test_does_not_update_user_if_lock_not_acquired` for an example of coordinating threads and opening another connection.

**Pull request merged**